nntrainer/tensor/tensor.cpp

```diff
@@ -80,18 +80,80 @@ static auto rng = [] {
   return rng;
 }();
 
-Tensor::Tensor(const TensorDim &d, const float *buf) : Tensor() {
+Tensor::Tensor(const TensorDim &d, const float *buf, bool alloc_now) : Tensor() {
   if (d.getDataLen() != 0) {
+    if (buf != nullptr && !alloc_now)
+      throw std::invalid_argument("alloc_now must be true to create a tensor with a buffer");
+
     dim = d;
     strides = d.computeStrides();
-    data = std::shared_ptr<float>(new float[d.getDataLen()],
-                                  std::default_delete<float[]>());
+
+    if (alloc_now)
+      data = std::shared_ptr<float>(new float[d.getDataLen()],
+          std::default_delete<float[]>());
 
     if (buf != nullptr)
       copy(buf);
   }
 }
 
+/**
+ * @class SrcSharedTensor
+ * @brief Source of the shared tensor
+ */
+class SrcSharedTensor {
+public:
+  /**
+   * @brief   Constructor for the class
+   */
+  SrcSharedTensor(): src(nullptr), off(0) {}
+
+  SrcSharedTensor(const Tensor *tensor, unsigned int offset): src(tensor), off(offset) {}
+
+  /**
+   * @brief   Get the allocated src tensor
+   */
+  const Tensor *tensor() {
+    if (!src)
+      throw std::runtime_error("Accessing empty src tensor");
+
+    /**
+     * Allocate source tensor if not already allocated.
+     */
+    if (!src->isAllocated())
+      throw std::runtime_error("Source tensor must be allocated before allocating dependent tensors");
+      // src->allocate();
+
+    return src;
+  }
+
+  /**
+   * @brief   Get the offset from the source tensor
+   */
+  unsigned int offset() {
+    return off;
+  }
+
+private:
+  const Tensor *src;   /**< Tensor of the source */
+  unsigned int off;  /**< offset from the source data ptr */
+};
+
+void Tensor::allocate() {
+  if (data)
+    /// already allocated
+    return;
+
+  if (src_tensor) {
+    /// allocate data based on the source tensor
+    data = std::shared_ptr<float>(src_tensor->tensor()->data, src_tensor->tensor()->data.get() + src_tensor->offset());
```

Comment:
how about hiding this logic into `src_tensor->allocate()`

Reply:
`allocate` is the only function which allocates/sets `data` in tensor.
I didn't want to split it.

Reply:
Sure